kvcoord: test additional ambiguous failure states

AlexTalks opened the pull request:
This change builds on the testing introduced in https://github.com/cockroachdb/cockroach/pull/107323 with additional
subtests evaluating the behavior of different race outcomes with
contended transactions where the first transaction experiences an RPC
failure (i.e. an ambiguous write).

Depends on #107323, #108154.

Part of: https://github.com/cockroachdb/cockroach/issues/103817

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/108001)
<!-- Reviewable:end -->

AlexTalks committed:
kvcoord: test additional ambiguous failure states

This change builds on the testing introduced in #107323 with additional
subtests evaluating the behavior of different race outcomes with
contended transactions where the first transaction experiences an RPC
failure (i.e. an ambiguous write).

Part of: #103817

Release note: None

arulajmani reviewed (commented):
Nice job with the thorough tests here!

It seems like there's a bit of copy pasting that's going on between each subtest. I'm wondering if we can get rid of some of it, as it'll make these tests much more readable. One opportunity, like I mentioned below, that comes to mind is to use a table driven test where we inject different "txn2"s for each subtest. That'll allow us to condense 3 variants (writer-reader conflict, writer-writer conflict, and workload) into one table driven test. 

Maybe there's something we can do for some of the other variants as well, by grouping some test cases together? What do you think?

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108001#-:-NcyCayL13Ff9LYQXRyN:b-l5u9w2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @AlexTalks)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender_ambiguous_test.go` line 558 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/108001#-Ncy78UwB3_ghMCw9IVb:-Ncy78UwB3_ghMCw9IVc:b-holyb3) ([raw file](https://github.com/cockroachdb/cockroach/blob/4836efc084af6550f8364c6968f9b66b7a658538/pkg/kv/kvclient/kvcoord/dist_sender_ambiguous_test.go#L558)):*
> ```Go
> 			tCtx := context.Background()
> 
> 			// txn2 just performs a simple GetForUpdate on a conflicting key, causing
> ```

Curious, is there a reason for the locking Get? Or would a non-locking Get have worked just as well? 
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_ambiguous_test.go` line 692 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/108001#-Ncy7eBwDnLtYEOUpiu_:-Ncy7eBwDnLtYEOUpiua:bjcqosb) ([raw file](https://github.com/cockroachdb/cockroach/blob/4836efc084af6550f8364c6968f9b66b7a658538/pkg/kv/kvclient/kvcoord/dist_sender_ambiguous_test.go#L692)):*
> ```Go
> 			tCtx := context.Background()
> 
> 			// txn2 performs simple Puts on conflicting keys, causing it to issue a
> ```

There's a bit of copy pasting between this subtest, the writer-reader subtest, and the workload conflict subtest. 

I think the only thing that's changing between these 3 subtests is txn2. Would it be possible to inject txn2 into each of these 3 subtests, instead of copy pasting the entire subtest 3 times? Maybe a table driven test like:

```go

  
testCases := []struct{  
 name string  
 txn2 func(wg sync.WaitGroup, txn2Ready chan struct{})  
}{  
 ...  
}

```
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_ambiguous_test.go` line 967 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/108001#-NcyB1c77LVXE4qVFSlf:-NcyB1c89ad4ePis6kIX:b-b6z45m) ([raw file](https://github.com/cockroachdb/cockroach/blob/4836efc084af6550f8364c6968f9b66b7a658538/pkg/kv/kvclient/kvcoord/dist_sender_ambiguous_test.go#L967)):*
> ```Go
> 			tCtx := context.Background()
> 			txn := db.NewTxn(tCtx, "txn1")
> 			vals := getInBatch(tCtx, txn, keyA, keyB, keyAPrime)
> ```

Could this test be fit into the table driven testCases approach if we made `txn1` a parameter as well?


<!-- Sent from Reviewable.io -->

AlexTalks reviewed (commented):
> Reviewed 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108001)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender_ambiguous_test.go` line 558 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/108001#-Ncy78UwB3_ghMCw9IVb:-Nd26K4TEmcUbax-NjlZ:b8jj4ya) ([raw file](https://github.com/cockroachdb/cockroach/blob/4836efc084af6550f8364c6968f9b66b7a658538/pkg/kv/kvclient/kvcoord/dist_sender_ambiguous_test.go#L558)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Curious, is there a reason for the locking Get? Or would a non-locking Get have worked just as well?
</blockquote></details>

Done - I don't think a `GetForUpdate` is required here, this was just so I could reuse the function that does the `GetForUpdate`s in a  batch, but we can certainly change this to be a simple `txn.Get`...
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_ambiguous_test.go` line 692 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/108001#-Ncy7eBwDnLtYEOUpiu_:-Nd2ahV3EOmLeNytBN9y:b-psy9f5) ([raw file](https://github.com/cockroachdb/cockroach/blob/4836efc084af6550f8364c6968f9b66b7a658538/pkg/kv/kvclient/kvcoord/dist_sender_ambiguous_test.go#L692)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

There's a bit of copy pasting between this subtest, the writer-reader subtest, and the workload conflict subtest. 

I think the only thing that's changing between these 3 subtests is txn2. Would it be possible to inject txn2 into each of these 3 subtests, instead of copy pasting the entire subtest 3 times? Maybe a table driven test like:

```go

  
testCases := []struct{  
 name string  
 txn2 func(wg sync.WaitGroup, txn2Ready chan struct{})  
}{  
 ...  
}

```
</blockquote></details>

Done - was able to do this, but only for the first three, which have identical concurrent operations (txn1/txn2/lease mover), and (nearly) identical request schedules.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_ambiguous_test.go` line 967 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/108001#-NcyB1c77LVXE4qVFSlf:-Nd2_ynu7UG5By6B5zrT:baz0ggq) ([raw file](https://github.com/cockroachdb/cockroach/blob/4836efc084af6550f8364c6968f9b66b7a658538/pkg/kv/kvclient/kvcoord/dist_sender_ambiguous_test.go#L967)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Could this test be fit into the table driven testCases approach if we made `txn1` a parameter as well?
</blockquote></details>

Unfortunately no, the schedule of intercepted requests is different in each of the cases apart from the first three, as are the signals/checkpoints in each test.

Keep in mind, the goal is to eventually transition to a data-driven test framework - this is meant to provide intermediary coverage for the immediately following fix PR.  So while it is important for these tests to be stable, not flakey, and cover the various scenarios possible, the framework itself should be considered a work in progress (and should hopefully not block the fix).  

All that is to say, I'm not sure rewriting the framework again now should be the priority.


<!-- Sent from Reviewable.io -->

AlexTalks reviewed (commented):
Refactored the ones with identical schedules, but I'm not sure it's reasonable for the other ones at this stage (see comments inline).

The `maybeWait` functions have some similarities, but each are distinct as each test examines the result of a slightly different operation schedule.  As such, the framework was rewritten in this way from earlier feedback that said following the request interception flow was too confusing when it was separated from the test itself, and it would be much easier to follow if we could trace the flow of operations in order for each test.

Hence, refactoring this to fit into a table driven test would require allowing declaration of generalized named checkpoints (ie channels), and defining a way to write the order of operations as scheduled in the `maybeWait` functions into the table-driven test - which may still end up making these things more confusing.  I'd like to defer further enhancements like this to our discussed work on a higher-level data-driven test framework, if that makes sense.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108001#-:-Nd2gC74AAEgspHge_Q-:b9d436d)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)



<!-- Sent from Reviewable.io -->

arulajmani reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> modulo comments. 

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108001#-:-NdBaiEzBncQUeSFT-5K:b-cm7ocf)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @AlexTalks)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender_ambiguous_test.go` line 692 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/108001#-Ncy7eBwDnLtYEOUpiu_:-NdAt11001p-XuQQGq2l:b-a93gji) ([raw file](https://github.com/cockroachdb/cockroach/blob/4836efc084af6550f8364c6968f9b66b7a658538/pkg/kv/kvclient/kvcoord/dist_sender_ambiguous_test.go#L692)):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

Done - was able to do this, but only for the first three, which have identical concurrent operations (txn1/txn2/lease mover), and (nearly) identical request schedules.
</blockquote></details>

Ack, thanks for pulling the ones we could here. It's a huge readability win :100: 
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_ambiguous_test.go` line 705 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/108001#-NdAuZZ28VpGoIRNufjL:-NdAuZZ28VpGoIRNufjM:b-1f850r) ([raw file](https://github.com/cockroachdb/cockroach/blob/69a94bf86d9f649fb0248f110998d16b4e154c88/pkg/kv/kvclient/kvcoord/dist_sender_ambiguous_test.go#L705)):*
> ```Go
> 	}
> 
> 	for _, variant := range basicVariants {
> ```

nit:  let's pull the `basicVariants` definition right above this loop.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_ambiguous_test.go` line 917 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/108001#-NdAwnrRDlhgI20Adu0-:-NdAwnrRDlhgI20Adu00:bjvb4em) ([raw file](https://github.com/cockroachdb/cockroach/blob/69a94bf86d9f649fb0248f110998d16b4e154c88/pkg/kv/kvclient/kvcoord/dist_sender_ambiguous_test.go#L917)):*
> ```Go
> 			}
> 
> 			// 5. txn3->n2: Put(c), EndTxn(parallel commit)
> ```

nit: I think txn3 should be able to hit 1PC,  so maybe `txn3->n2: 1PC`.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_ambiguous_test.go` line 998 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/108001#-NdAzzts7IfsEyQaOd2K:-NdAzzts7IfsEyQaOd2L:b-rcy279) ([raw file](https://github.com/cockroachdb/cockroach/blob/69a94bf86d9f649fb0248f110998d16b4e154c88/pkg/kv/kvclient/kvcoord/dist_sender_ambiguous_test.go#L998)):*
> ```Go
> 			"expected no AssertionFailedError due to sanity check")
> 	})
> 	t.Run("recovery after transfer lease", func(t *testing.T) {
> ```

Is there something special about this test case (over the transfer lease happens after recovery variants) or are we just adding it for completeness? Either way is fine, but maybe let's add a comment explaining why for future readers.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_ambiguous_test.go` line 1099 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/108001#-NdB14YDE6ulTOMBvByD:-NdB14YDE6ulTOMBvByE:b-srzi0e) ([raw file](https://github.com/cockroachdb/cockroach/blob/69a94bf86d9f649fb0248f110998d16b4e154c88/pkg/kv/kvclient/kvcoord/dist_sender_ambiguous_test.go#L1099)):*
> ```Go
> 			"expected AssertionFailedError due to sanity check on transaction already committed")
> 	})
> 	t.Run("retry sees other intent", func(t *testing.T) {
> ```

Similar to my comment above, a high level comment explaining why this test case is interesting would be helpful.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_ambiguous_test.go` line 1146 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/108001#-NdB_90wDXq5MAnEv2ak:-NdB_90wDXq5MAnEv2al:bht29u0) ([raw file](https://github.com/cockroachdb/cockroach/blob/69a94bf86d9f649fb0248f110998d16b4e154c88/pkg/kv/kvclient/kvcoord/dist_sender_ambiguous_test.go#L1146)):*
> ```Go
> 			// 9. txn2->n1: Put(a), EndTxn(parallel commit) -- Writes intent.
> 			// 10. txn2->n1: Put(b)
> 			// 11. txn2->n1: EndTxn(commit)
> ```

This is a bit subtle -- we only expect this to happen if the parallel commit attempt failed, but we were able to retry it. Should we add a comment explaining this?


<!-- Sent from Reviewable.io -->

AlexTalks reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108001)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @arulajmani)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender_ambiguous_test.go` line 917 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/108001#-NdAwnrRDlhgI20Adu0-:-NdCMVLD8gwltleCde43:bv9pu3f) ([raw file](https://github.com/cockroachdb/cockroach/blob/69a94bf86d9f649fb0248f110998d16b4e154c88/pkg/kv/kvclient/kvcoord/dist_sender_ambiguous_test.go#L917)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: I think txn3 should be able to hit 1PC,  so maybe `txn3->n2: 1PC`.
</blockquote></details>

I think you're right, good catch ;-)  Added comment
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_ambiguous_test.go` line 998 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/108001#-NdAzzts7IfsEyQaOd2K:-NdCMsH--zo6zf-W0UiW:b-9seot3) ([raw file](https://github.com/cockroachdb/cockroach/blob/69a94bf86d9f649fb0248f110998d16b4e154c88/pkg/kv/kvclient/kvcoord/dist_sender_ambiguous_test.go#L998)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Is there something special about this test case (over the transfer lease happens after recovery variants) or are we just adding it for completeness? Either way is fine, but maybe let's add a comment explaining why for future readers.
</blockquote></details>

This one was specifically because you asked about it actually, heh - I'll add a comment but if we don't think it's necessary I can remove it.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_ambiguous_test.go` line 1099 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/108001#-NdB14YDE6ulTOMBvByD:-NdCSpPe7Wz3EJlSbHsE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/69a94bf86d9f649fb0248f110998d16b4e154c88/pkg/kv/kvclient/kvcoord/dist_sender_ambiguous_test.go#L1099)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Similar to my comment above, a high level comment explaining why this test case is interesting would be helpful.
</blockquote></details>

Done.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_ambiguous_test.go` line 1146 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/108001#-NdB_90wDXq5MAnEv2ak:-NdCT1Ag7HkP5pIvN1w6:b-pth44x) ([raw file](https://github.com/cockroachdb/cockroach/blob/69a94bf86d9f649fb0248f110998d16b4e154c88/pkg/kv/kvclient/kvcoord/dist_sender_ambiguous_test.go#L1146)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

This is a bit subtle -- we only expect this to happen if the parallel commit attempt failed, but we were able to retry it. Should we add a comment explaining this?
</blockquote></details>

I don't think it's super relevant - even though it's asynchronous, it still happens - but I added a comment.


<!-- Sent from Reviewable.io -->

AlexTalks commented:
bors r+
